From patchwork Tue Jul  2 15:04:34 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Paolo Bonzini <pbonzini@redhat.com>
X-Patchwork-Id: 11028025
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id E12B7138D
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue,  2 Jul 2019 15:04:44 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C84362891F
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue,  2 Jul 2019 15:04:44 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id BBF6B2892B; Tue,  2 Jul 2019 15:04:44 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 64D8B2891F
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue,  2 Jul 2019 15:04:44 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727142AbfGBPEn (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 2 Jul 2019 11:04:43 -0400
Received: from mail-wm1-f65.google.com ([209.85.128.65]:36328 "EHLO
        mail-wm1-f65.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727063AbfGBPEk (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 2 Jul 2019 11:04:40 -0400
Received: by mail-wm1-f65.google.com with SMTP id u8so1422442wmm.1;
        Tue, 02 Jul 2019 08:04:40 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=sender:from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=qX4+tt0pDaQiw1/vCiXNQrZAln54Qf4UHvOvV7lW1a4=;
        b=GNx5m+StDBNK+z+0e4fWD4sLiRwhAKmGH7kkQS55kXOlhSH+sqqWVpivGYrIloeX6p
         rfbb0O38w/df7DVKwi0ySQLC1yHmxJOI57eSEPD4SrmvOMQFBT7z+S8oviZwzIVlqL3F
         crZ4ypGW7jfwLOmTXqXGj3P9CMuouyXrsiiD61wzs4nSdC7kg1HzwjZ+J4ObCfLZ7nT1
         D6F8TKBXJglkhf3n8FymISsDwrBFYXnxxSioeWXsNmm+kPkbthCPRab2sqMlP5wxqDUU
         9eSsD/NumGbgp2L+mCgyXmkZBVyBbysJui6NXHacrtYytSdyDzqHYUTX7MeVVDGqN2xI
         TkYQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:sender:from:to:cc:subject:date:message-id
         :in-reply-to:references;
        bh=qX4+tt0pDaQiw1/vCiXNQrZAln54Qf4UHvOvV7lW1a4=;
        b=SMPwu9YndHyHM1W1JwKaXlEgXKMMGFEl89Wi8sdcqQf0WzLinnrSmzBoAjb0UiAgsP
         JGBuBb5eiFNGvnpVcakWBG0qgzd7qsprN4Q43MCCSexaXLg15TXXS1YgwSyVchLA224x
         F4gTArsHuWDwea7MBKhjyPsKhT/NvMHW599eu+vf2yjPNo4sPwijwP1YcwZfqTYNB4Fw
         P+sNLGh/VIpSIeJjeQ2RdisQFQ7Zjrj/+biQCX8z31HIXbJLxD9Vzl2S3a8TlSsJaEe2
         sGWyHdxy5yYxAHtUQ4i4A6n3axa3GyAx2Jq4+DMtpSurvARLfwY7HoApj8foMta0ZGlW
         hNxg==
X-Gm-Message-State: APjAAAUbk/ibJiACRDKZzTCW/iaK/ALg9vknD8unpZCh7HY27P0DZk78
        ekHMzDwVY8q7ac2REQgPRdwum9tbR1M=
X-Google-Smtp-Source: 
 APXvYqwK6E1fV3qd3NM0eMoE80LVFbhqu0bmqdyCHPqY/qri01JijzdWRWX6hFwqnVrA+hwszI/pUA==
X-Received: by 2002:a05:600c:291:: with SMTP id
 17mr3702699wmk.32.1562079878935;
        Tue, 02 Jul 2019 08:04:38 -0700 (PDT)
Received: from 640k.lan ([93.56.166.5])
        by smtp.gmail.com with ESMTPSA id
 b203sm3494191wmd.41.2019.07.02.08.04.37
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 02 Jul 2019 08:04:38 -0700 (PDT)
From: Paolo Bonzini <pbonzini@redhat.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Liran Alon <liran.alon@oracle.com>
Subject: [PATCH 1/3] KVM: nVMX: include conditional controls in /dev/kvm
 KVM_GET_MSRS
Date: Tue,  2 Jul 2019 17:04:34 +0200
Message-Id: <1562079876-20756-2-git-send-email-pbonzini@redhat.com>
X-Mailer: git-send-email 1.8.3.1
In-Reply-To: <1562079876-20756-1-git-send-email-pbonzini@redhat.com>
References: <1562079876-20756-1-git-send-email-pbonzini@redhat.com>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Some secondary controls are automatically enabled/disabled based on the CPUID
values that are set for the guest.  However, they are still available at a
global level and therefore should be present when KVM_GET_MSRS is sent to
/dev/kvm.

Fixes: 1389309c811 ("KVM: nVMX: expose VMX capabilities for nested hypervisors to userspace", 2018-02-26)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
Reviewed-by: Liran Alon <liran.alon@oracle.com>
---
 arch/x86/kvm/vmx/nested.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/vmx/nested.c b/arch/x86/kvm/vmx/nested.c
index 990e543f4531..c4e29ef0b21e 100644
--- a/arch/x86/kvm/vmx/nested.c
+++ b/arch/x86/kvm/vmx/nested.c
@@ -5750,10 +5750,15 @@ void nested_vmx_setup_ctls_msrs(struct nested_vmx_msrs *msrs, u32 ept_caps,
 	msrs->secondary_ctls_low = 0;
 	msrs->secondary_ctls_high &=
 		SECONDARY_EXEC_DESC |
+		SECONDARY_EXEC_RDTSCP |
 		SECONDARY_EXEC_VIRTUALIZE_X2APIC_MODE |
+		SECONDARY_EXEC_WBINVD_EXITING |
 		SECONDARY_EXEC_APIC_REGISTER_VIRT |
 		SECONDARY_EXEC_VIRTUAL_INTR_DELIVERY |
-		SECONDARY_EXEC_WBINVD_EXITING;
+		SECONDARY_EXEC_RDRAND_EXITING |
+		SECONDARY_EXEC_ENABLE_INVPCID |
+		SECONDARY_EXEC_RDSEED_EXITING |
+		SECONDARY_EXEC_XSAVES;
 
 	/*
 	 * We can emulate "VMCS shadowing," even if the hardware

From patchwork Tue Jul  2 15:04:35 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Paolo Bonzini <pbonzini@redhat.com>
X-Patchwork-Id: 11028027
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 3672913B1
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue,  2 Jul 2019 15:04:56 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 1DF8F28751
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue,  2 Jul 2019 15:04:56 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 123672892B; Tue,  2 Jul 2019 15:04:56 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 8E38F2891F
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue,  2 Jul 2019 15:04:55 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727193AbfGBPEs (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 2 Jul 2019 11:04:48 -0400
Received: from mail-wm1-f67.google.com ([209.85.128.67]:39712 "EHLO
        mail-wm1-f67.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1725981AbfGBPEn (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 2 Jul 2019 11:04:43 -0400
Received: by mail-wm1-f67.google.com with SMTP id z23so1399682wma.4;
        Tue, 02 Jul 2019 08:04:41 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=sender:from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=ZRNHlXUigXjGU5a3lZWUqJMuxAc6p2T+k9d3I5VYScQ=;
        b=Tetkj5G0/Te2mIwrsStd90h3ChcvHTEZq0KzFhLk1LBcqcarl3rWKPN9x1KvidfSNi
         Mp5OCpWeGRCD5izJ/6U8AYI8CIkFbZvKBWgYefKeY456FrecoLo7pbcBcxToZs5xIg4x
         c8QgsvhGF6uEF6ZILMKlbzhassjEc8S60AsYg2E+YITt2ZhFi8jxgfuHYRZfN55rJd1t
         PlgOqHAW2NY7It1B7/+Wj8WEEtRZXDEr/QkzVVm5gwBdcu1ksiDgvQwiOZ1itaG2xumR
         swvjMIP27aMttemB8V9WoLQReScLMmBEcj9Bwzf8Veyk0PyKl76ptbdEewPW2Gt1iMS7
         qy3A==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:sender:from:to:cc:subject:date:message-id
         :in-reply-to:references;
        bh=ZRNHlXUigXjGU5a3lZWUqJMuxAc6p2T+k9d3I5VYScQ=;
        b=blRd2Usyqq0AqpW5uIxd/SGnjeX9rl/PrClRI9FcvnsNtKlsrd0LtI/PIZs52gXRs3
         kwhTPYLyt0RKeib2bu6Z7Fz9tNa9vZwSTs38hKeG4VU5LeCEMOQAsd+4lWveydcqeU3f
         T+XMN4QzPSvGbimrCQ0/HdEXf4n40xyNjv9RwuiJo5gu4T/HyHHrZf7BiSB1uUpeHmAb
         DbyBSwy3Ete+sVnAVT6kSvulSxx4zH9UBDd9PbSKtLrp5VPCay8bFAapHbNQEk+J2n2f
         mOrnuQJxXxfmmLOyysGbmZbL7PH4b0oPkWvWMc6XO+67jqAYZw5ANyVgSV9V9MHNj7VM
         iN4A==
X-Gm-Message-State: APjAAAXXQ0H/Iiwc540OJaPg5mBbYV9J67YUfjRCIlB16e71Uf8uZhE9
        rK6GX2zJJXyidEF2u6pKQAFQ0BHCqno=
X-Google-Smtp-Source: 
 APXvYqxVsiR+PKXuwyeeuUHR/y+wM9C1MpcG3pmuOVHiatR1t0bY2kAeqgOE3d9q0Y9zA0ES9LQjpw==
X-Received: by 2002:a05:600c:204c:: with SMTP id
 p12mr3751294wmg.121.1562079880395;
        Tue, 02 Jul 2019 08:04:40 -0700 (PDT)
Received: from 640k.lan ([93.56.166.5])
        by smtp.gmail.com with ESMTPSA id
 b203sm3494191wmd.41.2019.07.02.08.04.38
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 02 Jul 2019 08:04:39 -0700 (PDT)
From: Paolo Bonzini <pbonzini@redhat.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Liran Alon <liran.alon@oracle.com>, stable@vger.kernel.org
Subject: [PATCH 2/3] KVM: nVMX: allow setting the VMFUNC controls MSR
Date: Tue,  2 Jul 2019 17:04:35 +0200
Message-Id: <1562079876-20756-3-git-send-email-pbonzini@redhat.com>
X-Mailer: git-send-email 1.8.3.1
In-Reply-To: <1562079876-20756-1-git-send-email-pbonzini@redhat.com>
References: <1562079876-20756-1-git-send-email-pbonzini@redhat.com>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Allow userspace to set a custom value for the VMFUNC controls MSR, as long
as the capabilities it advertises do not exceed those of the host.

Fixes: 27c42a1bb ("KVM: nVMX: Enable VMFUNC for the L1 hypervisor", 2017-08-03)
Cc: stable@vger.kernel.org
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
Reviewed-by: Liran Alon <liran.alon@oracle.com>
---
 arch/x86/kvm/vmx/nested.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/x86/kvm/vmx/nested.c b/arch/x86/kvm/vmx/nested.c
index c4e29ef0b21e..163d226efa96 100644
--- a/arch/x86/kvm/vmx/nested.c
+++ b/arch/x86/kvm/vmx/nested.c
@@ -1234,6 +1234,11 @@ int vmx_set_vmx_msr(struct kvm_vcpu *vcpu, u32 msr_index, u64 data)
 	case MSR_IA32_VMX_VMCS_ENUM:
 		vmx->nested.msrs.vmcs_enum = data;
 		return 0;
+	case MSR_IA32_VMX_VMFUNC:
+		if (data & ~vmx->nested.msrs.vmfunc_controls)
+			return -EINVAL;
+		vmx->nested.msrs.vmfunc_controls = data;
+		return 0;
 	default:
 		/*
 		 * The rest of the VMX capability MSRs do not support restore.

From patchwork Tue Jul  2 15:04:36 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Paolo Bonzini <pbonzini@redhat.com>
X-Patchwork-Id: 11028029
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 0F440138D
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue,  2 Jul 2019 15:04:58 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id EBA3828751
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue,  2 Jul 2019 15:04:57 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id E03C02892B; Tue,  2 Jul 2019 15:04:57 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 834EC28751
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue,  2 Jul 2019 15:04:57 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727179AbfGBPEs (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 2 Jul 2019 11:04:48 -0400
Received: from mail-wr1-f66.google.com ([209.85.221.66]:42067 "EHLO
        mail-wr1-f66.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727130AbfGBPEn (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 2 Jul 2019 11:04:43 -0400
Received: by mail-wr1-f66.google.com with SMTP id x17so18219705wrl.9;
        Tue, 02 Jul 2019 08:04:42 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=sender:from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=t56wQhmbC0eK6Si58Q9wqbieCBfEkE9Pa30U73bNL2s=;
        b=LPtQII5df23fZnUhJhsdshpakbDmVhjyV7jWcOKOIs2e/ZHfr3fEGBH+6nvj8DsUCY
         gLD40CWql4WreFhw2fHLTNvEji3Gr/jZMEm6Ix2brFwD5ml31Qy6rQDkTGs6RHhQnM4l
         mUWfYlVgBPJkrrCLfbj7AuCqPAmdaQ+lgGZGMaF4RAK9JciT3l4oNA/pNxbALGatPI++
         mJMPn864+cvRqJDTc5l9iWxqD/su98ET8kXVMtPpIWQzHLKbQPxcNgyTGlNCR9QpXO9J
         wzAaScV8Qjj7g7guI9XzAkBxRewlLFC5xgVrcDBmsU5Lma2DAVyB8wtTXrSfYpaGEhtW
         WlIQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:sender:from:to:cc:subject:date:message-id
         :in-reply-to:references;
        bh=t56wQhmbC0eK6Si58Q9wqbieCBfEkE9Pa30U73bNL2s=;
        b=qyfi0xxGQF6uxRqPIPVG/3Eai+ylZCuSKI0Ea+8TcvRfxfmabpKuJtlku5grvbeuBd
         +cBCKPdhjxzAX82i/vn77pXHNeSx00BuuCfai2/U4HKB5g0Kw4tC6cY7qY5YB4DAiAB6
         dxswabXF4qmVdMPms0nG631L1o1u8FGY39+GjyIwRwoAAQXjS3bRh1PjQj9C0iI7qTF+
         ZU31EBz13ueRMckIy4cwLcEmwxoSLZxGP6LFT1IqfDoZvZU5sbgGPSbRfaTSOIwju1B0
         z+Lehbpecm4FenamCyBAbfp627OQTjYNdNCkMXa5FqiUpnugPXxfwdCU/P3G/c+wLPgr
         A3yA==
X-Gm-Message-State: APjAAAWkdIzMvxxSNC8//7Ll8vRswSX5YyvyAh3gUiv+zumPHJg9agFQ
        Izhd4lT1bXHbkmX1HtxmUL+uTJJH+/w=
X-Google-Smtp-Source: 
 APXvYqwxRLKO+D0At4TubFsnDlNCAevYaJZOFetfxOe9d+M5+sj+wf62oBSmxnWtg7uOj+sXiB3/9g==
X-Received: by 2002:a5d:53ca:: with SMTP id
 a10mr24763517wrw.131.1562079881552;
        Tue, 02 Jul 2019 08:04:41 -0700 (PDT)
Received: from 640k.lan ([93.56.166.5])
        by smtp.gmail.com with ESMTPSA id
 b203sm3494191wmd.41.2019.07.02.08.04.40
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 02 Jul 2019 08:04:40 -0700 (PDT)
From: Paolo Bonzini <pbonzini@redhat.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Liran Alon <liran.alon@oracle.com>, stable@vger.kernel.org
Subject: [PATCH 3/3] KVM: nVMX: list VMX MSRs in KVM_GET_MSR_INDEX_LIST
Date: Tue,  2 Jul 2019 17:04:36 +0200
Message-Id: <1562079876-20756-4-git-send-email-pbonzini@redhat.com>
X-Mailer: git-send-email 1.8.3.1
In-Reply-To: <1562079876-20756-1-git-send-email-pbonzini@redhat.com>
References: <1562079876-20756-1-git-send-email-pbonzini@redhat.com>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

This allows userspace to know which MSRs are supported by the hypervisor.
Unfortunately userspace must resort to tricks for everything except
MSR_IA32_VMX_VMFUNC (which was just added in the previous patch).
One possibility is to use the feature control MSR, which is tied to nested
VMX as well and is present on all KVM versions that support feature MSRs.

Fixes: 1389309c811 ("KVM: nVMX: expose VMX capabilities for nested hypervisors to userspace", 2018-02-26)
Cc: stable@vger.kernel.org
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---
 arch/x86/kvm/svm.c     |  1 +
 arch/x86/kvm/vmx/vmx.c |  2 ++
 arch/x86/kvm/x86.c     | 20 ++++++++++++++++++++
 3 files changed, 23 insertions(+)

diff --git a/arch/x86/kvm/svm.c b/arch/x86/kvm/svm.c
index bbc31f7213ed..5db50c19d1c7 100644
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -5885,6 +5885,7 @@ static bool svm_has_emulated_msr(int index)
 {
 	switch (index) {
 	case MSR_IA32_MCG_EXT_CTL:
+	case MSR_IA32_VMX_BASIC ... MSR_IA32_VMX_VMFUNC:
 		return false;
 	default:
 		break;
diff --git a/arch/x86/kvm/vmx/vmx.c b/arch/x86/kvm/vmx/vmx.c
index a35459ce7e29..c43635942693 100644
--- a/arch/x86/kvm/vmx/vmx.c
+++ b/arch/x86/kvm/vmx/vmx.c
@@ -6223,6 +6223,8 @@ static bool vmx_has_emulated_msr(int index)
 		 * real mode.
 		 */
 		return enable_unrestricted_guest || emulate_invalid_guest_state;
+	case MSR_IA32_VMX_BASIC ... MSR_IA32_VMX_VMFUNC:
+		return nested;
 	case MSR_AMD64_VIRT_SPEC_CTRL:
 		/* This is AMD only.  */
 		return false;
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 8996a3131116..a02d4c244422 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -1177,6 +1177,26 @@ bool kvm_rdpmc(struct kvm_vcpu *vcpu)
 	MSR_AMD64_VIRT_SPEC_CTRL,
 	MSR_IA32_POWER_CTL,
 
+	/*
+	 * The following list leaves out MSRs whose values are determined
+	 * by arch/x86/kvm/vmx/nested.c based on CPUID or other MSRs.
+	 * We always support the "true" VMX control MSRs, even if the host
+	 * processor does not, so I am putting these registers here rather
+	 * than in msrs_to_save.
+	 */
+	MSR_IA32_VMX_BASIC,
+	MSR_IA32_VMX_TRUE_PINBASED_CTLS,
+	MSR_IA32_VMX_TRUE_PROCBASED_CTLS,
+	MSR_IA32_VMX_TRUE_EXIT_CTLS,
+	MSR_IA32_VMX_TRUE_ENTRY_CTLS,
+	MSR_IA32_VMX_MISC,
+	MSR_IA32_VMX_CR0_FIXED0,
+	MSR_IA32_VMX_CR4_FIXED0,
+	MSR_IA32_VMX_VMCS_ENUM,
+	MSR_IA32_VMX_PROCBASED_CTLS2,
+	MSR_IA32_VMX_EPT_VPID_CAP,
+	MSR_IA32_VMX_VMFUNC,
+
 	MSR_K7_HWCR,
 	MSR_KVM_POLL_CONTROL,
 };
